plugins/woocommerce/includes/shipping/flat-rate/class-wc-shipping-flat-rate.php

```diff
@@ -173,11 +173,41 @@ public function calculate_shipping( $package = array() ) {
 			$found_shipping_classes = $this->find_shipping_classes( $package );
 			$highest_class_cost     = 0;
 
+			// Store products to be shipped with every shipping class.
+			$shippable_products = array();
+
+			// Flag for indicating whether or not we are able to ship the entire cart with a single class.
+			$found_product_not_shippable_with_all_classes = false;
+
+			// Store this is the first iteration of shipping classes loop.
+			$first_iteration = true;
+
 			foreach ( $found_shipping_classes as $shipping_class => $products ) {
 				// Also handles BW compatibility when slugs were used instead of ids.
 				$shipping_class_term = get_term_by( 'slug', $shipping_class, 'product_shipping_class' );
 				$class_cost_string   = $shipping_class_term && $shipping_class_term->term_id ? $this->get_option( 'class_cost_' . $shipping_class_term->term_id, $this->get_option( 'class_cost_' . $shipping_class, '' ) ) : $this->get_option( 'no_class_cost', '' );
 
+				foreach ( $products as $data ) {
```

Comment:
I'm wondering if we can avoid this inner loop 🤔 

Looking at the [current implementation](https://github.com/woocommerce/woocommerce/blob/7.7.0/plugins/woocommerce/includes/shipping/flat-rate/class-wc-shipping-flat-rate.php#L146-L217), we only add a rate if we've determined that [there are shipping costs](https://github.com/woocommerce/woocommerce/blob/7.7.0/plugins/woocommerce/includes/shipping/flat-rate/class-wc-shipping-flat-rate.php#L206-L208) and, for that [to be true](https://github.com/woocommerce/woocommerce/blob/7.7.0/plugins/woocommerce/includes/shipping/flat-rate/class-wc-shipping-flat-rate.php#L185), we must have discovered a shipping class that yields a [non-empty class cost string](https://github.com/woocommerce/woocommerce/blob/7.7.0/plugins/woocommerce/includes/shipping/flat-rate/class-wc-shipping-flat-rate.php#L181).

If the class cost string is empty, **and** if the calculation type is per class (ie, `'class' === $this->type`), perhaps that is all the reason we need to just bail out of the `$found_shipping_classes` loop? Something like this:

```php
$no_class_cost = 0 === strlen( $class_cost_string );

if ( $no_class_cost && 'class' === $this->type ) {
	// If we are calculating per each individual shipping class, but one of the products belongs to a
	// class for which a cost has not been defined, then we cannot supply a shipping rate.
	$has_costs = false;
	break;
} elseif ( $no_class_cost ) {
	// Otherwise, if we do not have a class cost but we are intending to use the highest available
	// rate, we should simply skip this product.
	continue;
}
```

If this is viable, it would be more efficient (since we don't need to loop through all of the products).

Reply:
Yes @barryhughes you were right. Missed this check. Checked it out and it works great and its definitely more performant.